src/map/achievement.cpp

```diff
@@ -1197,83 +1200,31 @@ void achievement_script_free(struct av_condition *condition)
 	aFree(condition);
 }
 
-/**
- * Clear achievement single entry
- * @param achievement: Achievement to clear
- * @param free: Will free achievement from memory
- */
-void achievement_db_free_sub(struct achievement_db *achievement, bool free)
-{
-	if (achievement->targets) {
-		aFree(achievement->targets);
-		achievement->targets = NULL;
-		achievement->target_count = 0;
-	}
-	if (achievement->condition) {
-		achievement_script_free(achievement->condition);
-		achievement->condition = NULL;
-	}
-	if (achievement->dependents) {
-		aFree(achievement->dependents);
-		achievement->dependents = NULL;
-		achievement->dependent_count = 0;
-	}
-	if (achievement->rewards.script) {
-		script_free_code(achievement->rewards.script);
-		achievement->rewards.script = NULL;
-	}
-	if (free)
-		aFree(achievement);
-}
-
-/**
- * Clears the achievement database for shutdown or reload.
- */
-static int achievement_db_free(DBKey key, DBData *data, va_list ap)
-{
-	struct achievement_db *achievement = (struct achievement_db *)db_data2ptr(data);
-
-	if (!achievement)
-		return 0;
-
-	achievement_db_free_sub(achievement, true);
-	return 1;
-}
-
-static int achievementmobs_db_free(DBKey key, DBData *data, va_list ap)
-{
-	struct achievementmobs_db *achievement = (struct achievementmobs_db *)db_data2ptr(data);
-
-	if (!achievement)
-		return 0;
-
-	aFree(achievement);
-	return 1;
-}
-
 void achievement_db_reload(void)
 {
 	if (!battle_config.feature_achievement)
 		return;
-	achievementmobs_db->clear(achievementmobs_db, achievementmobs_db_free);
-	achievement_db->clear(achievement_db, achievement_db_free);
-	achievement_read_db();
+	do_final_achievement();
+	do_init_achievement();
 }
 
 void do_init_achievement(void)
 {
 	if (!battle_config.feature_achievement)
 		return;
-	memset(&achievement_dummy, 0, sizeof(achievement_dummy));
-	achievement_db = idb_alloc(DB_OPT_BASE);
-	achievementmobs_db = idb_alloc(DB_OPT_BASE);
 	achievement_read_db();
 }
 
 void do_final_achievement(void)
 {
 	if (!battle_config.feature_achievement)
```

Comment:
I have a feeling that check could cause some leak. (Well no more with your change but I just feel it's error prone)